### PR TITLE
Mock ocall in sgx_pthread.edl for in-enclave RA verifier

### PR DESCRIPTION
**Pull request by duanbing**
Per as https://github.com/intel/SGXDataCenterAttestationPrimitives/pull/193,  some thread related `ocall` were added.  we here mock those APIs for graphene `ra-tls`.
## How to test this PR? 

---
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)

### CODE REVIEW

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @duanbing)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/65#-MjhiwmE0hwAFvb6DSjC:-MjhiwmF3vaqLW1pd442:b-ubjt94) (no related file):*
FYI: For other reviewers: this is a legit issue. Basically, in some cases we want to run the RA-TLS verifier code inside the SGX enclave (though typically the verifier runs on a trusted machine and thus doesn't need SGX). To run the RA-TLS verifier, we used the Intel SGX SDK/PSW provided DCAP library. This DCAP library is not very well designed (actually, this was fixed recently in August, but didn't yet find its way to the commonly used and installed DCAP libs), so it mixes several unrelated functionalities. Some of these functionalities include doing pthreads-related OCALLs, which are Intel SGX SDK specific and obviously don't work in Graphene. Note that for quote verification code in this DCAP library, there is no need to get out of the enclave and thus no need in these OCALLs (so it works for our use case).

We already circumvented this problem once, about 6 months ago (see the top comment in the changed file `ra_tls_verify_dcap_urts.c`). But the DCAP lib added some more OCALLs in these 6 months, so we need to mock them as well.

I will probably add this explanation to the commit message body, so it makes sense to future readers.

---
___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/65#-MjhigXG1Gr_knZOi9YZ):*
You forgot to add the "Signed-off-by" line to your commit. Also, the prefix the commit title should be `[Pal/Linux-SGX]`.

But don't do it: I will do it myself during final rebase.

**Review by dimakuv — COMMENTED** [review]
@duanbing Thanks! This is correct now :)

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65#-:-Mjhk0XXDWb43meakb9s:bl2hqnm)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @duanbing)

**Review by mkow — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/65#-MjhigXG1Gr_knZOi9YZ):*
> But don't do it: I will do it myself during final rebase.

But formally you can't, the sign-off has to be added by the author, you can't sign the commit for him ;)

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @duanbing)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/65#-MjhigXG1Gr_knZOi9YZ):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

> But don't do it: I will do it myself during final rebase.

But formally you can't, the sign-off has to be added by the author, you can't sign the commit for him ;)
</blockquote></details>

@mkow Do you want @duanbing to do git-rebase then, to add this Sign-off message?

**Review by mkow — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @duanbing)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/65#-MjhigXG1Gr_knZOi9YZ):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@mkow Do you want @duanbing to do git-rebase then, to add this Sign-off message?
</blockquote></details>

The only other option would be that one of us just reimplement this fix and submit another PR. But we can't add the sign-off ourselves.

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @duanbing)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/65#-MjhigXG1Gr_knZOi9YZ):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

The only other option would be that one of us just reimplement this fix and submit another PR. But we can't add the sign-off ourselves.
</blockquote></details>

Fair enough. @duanbing Please use git-amend (or git-rebase, if you want) to add the Sign-off line in your commit message and then do git-push force.

**Review by duanbing — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/65#-MjhigXG1Gr_knZOi9YZ):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Fair enough. @duanbing Please use git-amend (or git-rebase, if you want) to add the Sign-off line in your commit message and then do git-push force.
</blockquote></details>

Amened and added the prefix [Pal/Linux-SGX].  Please take a look again.  Thanks!

**Review by dimakuv — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/65#-MjmYBYc3ZwJY4ZgOFxi:-MjmYBYc3ZwJY4ZgOFxj:b-fjhe1r) (no related file):*
I will modify the commit message to have more explanations during the final rebase. But in general, approving this PR.

---

**Comment by dimakuv** [comment]
Jenkins, test this please

**Review by mkow — DISMISSED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required) (waiting on @duanbing)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/65#-MjhigXG1Gr_knZOi9YZ):*
<details><summary><i>Previously, duanbing (Bing) wrote…</i></summary><blockquote>

Amened and added the prefix [Pal/Linux-SGX].  Please take a look again.  Thanks!
</blockquote></details>

There's a missing space after `]`. But I'll leave this to @dimakuv to fix.

**Review by boryspoplawski — DISMISSED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 1 unresolved discussion (waiting on @mkow)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/65#-MjmYBYc3ZwJY4ZgOFxi:-MjoBnllCO1C4u_7Myt2:bbh6j1) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I will modify the commit message to have more explanations during the final rebase. But in general, approving this PR.
</blockquote></details>

Also we don't have `sgx_pthread.edl` so it sounds weird (I know it's file from linux-sgx)

---

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/65#-MjmYBYc3ZwJY4ZgOFxi:-MjoFs3yCCQeENmYk01y:b-l7u3n0) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Also we don't have `sgx_pthread.edl` so it sounds weird (I know it's file from linux-sgx)
</blockquote></details>

Done. Hopefully now it's all clear.

---

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required)

**Comment by mkow** [comment]
Jenkins, test this please

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required) (waiting on @duanbing)

___
*[Pal/src/host/Linux-SGX/tools/ra-tls/ra_tls_verify_dcap_urts.c, line 20 at r3](https://reviewable.io/reviews/gramineproject/gramine/65#-MjoGEcD6zgHv1KkZ44m:-MjoGEcD6zgHv1KkZ44n:b7ge1cr) ([raw file](https://github.com/gramineproject/gramine/blob/2d87d9063e72e7be68ba7e6da6232bea8b3dafbf/Pal/src/host/Linux-SGX/tools/ra-tls/ra_tls_verify_dcap_urts.c#L20)):*
> ```C
> DUMMY_FUNCTION(sgx_destroy_enclave)
> DUMMY_FUNCTION(sgx_ecall)
> /* see https://github.com/intel/linux-sgx/blob/master/common/inc/sgx_tstdc.edl */
> ```

Actually this should be link to specific commit, not `master`.
___
*[Pal/src/host/Linux-SGX/tools/ra-tls/ra_tls_verify_dcap_urts.c, line 26 at r3](https://reviewable.io/reviews/gramineproject/gramine/65#-MjoGIjsENK7Ljf9vP_J:-MjoGIjsENK7Ljf9vP_K:b-7ropln) ([raw file](https://github.com/gramineproject/gramine/blob/2d87d9063e72e7be68ba7e6da6232bea8b3dafbf/Pal/src/host/Linux-SGX/tools/ra-tls/ra_tls_verify_dcap_urts.c#L26)):*
> ```C
> DUMMY_FUNCTION(sgx_thread_set_multiple_untrusted_events_ocall)
> DUMMY_FUNCTION(sgx_thread_wait_untrusted_event_ocall)
> /* see https://github.com/intel/linux-sgx/blob/master/common/inc/sgx_pthread.edl */
> ```

ditto

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @boryspoplawski)

___
*[Pal/src/host/Linux-SGX/tools/ra-tls/ra_tls_verify_dcap_urts.c, line 20 at r3](https://reviewable.io/reviews/gramineproject/gramine/65#-MjoGEcD6zgHv1KkZ44m:-MjoGvxbDr-J2-f9H4tW:b3md9np) ([raw file](https://github.com/gramineproject/gramine/blob/2d87d9063e72e7be68ba7e6da6232bea8b3dafbf/Pal/src/host/Linux-SGX/tools/ra-tls/ra_tls_verify_dcap_urts.c#L20)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Actually this should be link to specific commit, not `master`.
</blockquote></details>

Done (used the latest tag)
___
*[Pal/src/host/Linux-SGX/tools/ra-tls/ra_tls_verify_dcap_urts.c, line 26 at r3](https://reviewable.io/reviews/gramineproject/gramine/65#-MjoGIjsENK7Ljf9vP_J:-MjoGwWg61lArLvoSQ6v:b-896fjb) ([raw file](https://github.com/gramineproject/gramine/blob/2d87d9063e72e7be68ba7e6da6232bea8b3dafbf/Pal/src/host/Linux-SGX/tools/ra-tls/ra_tls_verify_dcap_urts.c#L26)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done

**Review by boryspoplawski — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required)

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/65)* status: :shipit: complete! all files reviewed, all discussions resolved

**Comment by mkow** [comment]
Jenkins, retest this please

**Pull request merged**
